### PR TITLE
Rating Component

### PR DESCRIPTION
**Pull request by uranussg**

### CODE REVIEW

File: src/components/Rating/Rating.tsx [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+export interface IRatingProps {      
+    /** set customized style */
+    className?: string;
+    /** set icon shape */
+    rtShape?: RatingShape;
+    /** set icon size */
+    rtSize?: RatingSize;
+    /** set initial score */
+    rtInitScore?: number | (()=> Promise<any>);
```
Comment (YuhanLin1105) [inline src/components/Rating/Rating.tsx:47]:
```ts
rtInitScore?: number | (()=> Promise<number>);
```

File: src/components/Rating/Rating.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+);
+
+export const DiffOnSelectRating = () => (
+    <div style = {RatingStyle}>
+    <div className='rt-title'>ConsoleLog</div>
+    <Rating rtOnSelect={mockOnSLog} />
+    <div className='rt-title'>AjaxCall</div>
+    <Rating  rtOnSelect={mockOnSAjax}/>
+    <div className='rt-title'>API</div>
+    <Rating  rtOnSelect={mockOnSAPI} />
```
Comment (YuhanLin1105) [inline src/components/Rating/Rating.stories.tsx:101]:
```suggestion
    <Rating  rtOnSelect={(num)=>{action(num)()}} />
```
One example is enough Or removes `DiffOnSelectRating` and put  `rtOnSelect={(num)=>{action(num)()}} ` into other examples

File: public/index.html [hunk 3]
```diff
@@ -2,7 +2,7 @@
 <html lang="en">
   <head>
     <meta charset="utf-8" />
-    <link rel="icon" href="%PUBLIC_URL%/favicon.ico" />
+    <link rel="icon" href="%PUBLIC_URL%/star.ico" />
```
[the remark this answers is not in the document]
Comment (uranussg) [reply public/index.html]: I changed it back

File: src/components/Rating/Rating.tsx [hunk 4]
```diff
@@ -0,0 +1,138 @@
+import React, {FunctionComponent, MouseEvent, useEffect, useState} from 'react';
+// import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
+// import { IconProp } from '@fortawesome/fontawesome-svg-core'
+// import { library } from '@fortawesome/fontawesome-svg-core'
+// import { faMoon, faStar, faHeart, faSmileWink } from '@fortawesome/free-regular-svg-icons'
+// import {fas} from '@fortawesome/free-solid-svg-icons'
+// library.add( fas, faMoon, faStar, faHeart, faSmileWink )
```
[the remark this answers is not in the document]
Comment (uranussg) [reply src/components/Rating/Rating.tsx]: removed

File: src/components/Rating/Rating.tsx [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+//     Fade = 'fade',
+//     Bounce = 'bounce',
+//     Swing = 'swing',
+
+// }
+
+
+export interface IRatingProps {      
+    /** set unique rtKey */
+    rtKey: string | number
```
[the remark this answers is not in the document]
Comment (uranussg) [reply src/components/Rating/Rating.tsx]: changed to useRef as suggested

File: src/components/Rating/Rating.tsx [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    /** set unique rtKey */
+    rtKey: string | number 
+    /** set customized style */
+    className?: string;
+    /** set icon shape */
+    rtShape?: RatingShape;
+    /** set icon size */
+    rtSize?: RatingSize;
+    /** set initial score */
+    rtInitScore?:number | Function;
```
[the remark this answers is not in the document]
Comment (uranussg) [reply src/components/Rating/Rating.tsx]: changed to rtInitScore?: number | (()=> Promise<number>);

File: src/components/Rating/Rating.tsx [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    let list = (new Array(rtMaxScore).fill(0)).map((el, idx)=> {
+        return (<div className='rt-icon' onClick={handleClick} id={`rt-${rtMaxScore as number - idx}`} key = {idx}>
+                    {/* <FontAwesomeIcon icon={['fas', rtShape as string] as IconProp} size={rtSize}/> */}
+                    <Icon name={rtShape as string} size ={rtSize}/>
+                </div>)
+    })
+    return (
+   
+        <>
+            <div id={`${rtKey}`} className={styleClasses + ' rt-container'} {...(rest as IRatingProps)}>
```
[the remark this answers is not in the document]
Comment (uranussg) [reply src/components/Rating/Rating.tsx]: changed to useRef as suggested

File: src/components/Rating/Rating.tsx [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    /** set icon shape */
+    rtShape?: RatingShape;
+    /** set icon size */
+    rtSize?: RatingSize;
+    /** set initial score */
+    rtInitScore?:number | Function;
+    /** set max score */
+    rtMaxScore?:number;
+    /** set customiized function */
+    rtFunction?: (score:number) => any;
```
[the remark this answers is not in the document]
Comment (uranussg) [reply src/components/Rating/Rating.tsx]: renamed and add one example in the default part

File: src/components/Icon/Icons.tsx [hunk 9]
```diff
@@ -55,7 +55,7 @@ export const IconPath: IIconPath = {
       'M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z',
     viewBox: '0 0 512 512',
   },
-  ['smile wink']: {
+  ['smile-wink']: {
```
[the remark this answers is not in the document]
Comment (uranussg) [reply src/components/Icon/Icons.tsx]: changed it back

File: src/components/Rating/Rating.tsx [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        if (typeof rtInitScore === 'function') {
+            rtInitScore().then((data:number)=> {
+                setScore(data)})            
+        }
+    }, [])       
+    const prefix = new Array(5).fill('fas')
+    let styleClasses = ['rt', `rt-${rtSize}`, `rt-${rtShape}`, `rt-rtKey-${rtKey}`].join(' ')
+    if (className) styleClasses += ' ' + className
+    useEffect(()=> {
+        let parent = document.getElementById(`${rtKey}`)
```
[the remark this answers is not in the document]
Comment (uranussg) [reply src/components/Rating/Rating.tsx]: used useRef

File: src/components/Rating/Rating.tsx (hunk 1 above)
Comment (uranussg) [reply src/components/Rating/Rating.tsx]: changed as suggested

File: src/components/Rating/Rating.stories.tsx (hunk 2 above)
Comment (uranussg) [reply src/components/Rating/Rating.stories.tsx]: put one example in the default section

**Comment by dlyling** [comment]
checks passed and no conflicts, looks good!

File: src/components/Rating/Rating.tsx [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    let {className, rtShape, rtSize, rtAnimation, rtInitScore, rtMaxScore, rtOnSelect, children, ...rest} = props
+    let myRef = useRef<HTMLDivElement>()
+    const [score, setScore] = useState(typeof rtInitScore === 'number'? rtInitScore : 0)
+    useEffect(()=> {
+        if (typeof rtInitScore === 'function') {
+            rtInitScore().then((data:number)=> {
+                setScore(data)})            
+            }
+        }, [])       
+    const prefix = new Array(5).fill('fas')
```
[the remark this answers is not in the document]
Comment (uranussg) [reply src/components/Rating/Rating.tsx:76]: removed

File: src/components/Rating/Rating.tsx [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+            target.classList.remove(`rt-${rtAnimation}`)
+        }, 1000)
+        // let num = parseInt(target.id.split('-')[1])
+        setScore(parseInt(num))    
+    }
+
+
+    let list = (new Array(rtMaxScore).fill(0)).map((el, idx)=> {
+        return (<div className='rt-icon' onClick={handleClick(`${rtMaxScore as number - idx}`)}>
+                    {/* <FontAwesomeIcon icon={['fas', rtShape as string] as IconProp} size={rtSize}/> */}
```
[the remark this answers is not in the document]
Comment (uranussg) [reply src/components/Rating/Rating.tsx:107]: removed

File: src/App.tsx [hunk 13]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import Rating from './components/Rating/Rating'
+import Button from './components/Button/Button';
+
+function App() {
```
[the remark this answers is not in the document]
Comment (uranussg) [reply src/App.tsx:5]: removed

**Pull request closed without merging**